Add annotation to normalize empty directories in inputs files

ghale opened the pull request:
This adds a new annotation (`@IgnoreDirectories`) and a corresponding runtime API to allow file inputs to normalize out empty directories during up-to-date and build cache snapshotting.

pioterj commented:
Wouldn't it communicate the intent better if the annotation was named @IgnoreEmptyDirectories rather than @IgnoreDirectories? When I read ignored directory I think of all the directory being ignored too.

ghale commented:
@pioterj It was originally called `@IgnoreEmptyDirectories`, but @wolfs pointed out that if we ignore only empty directories then we'd still have problems with directories that are not empty, but contain only other directories that are empty.  The best way to deal with this is to ignore _all_ directories - and that's what `@IgnoreDirectories` does - it only snapshots files.  The path to the file is part of the file snapshot, so we won't miss things like directories being renamed or anything (unless they are empty).  The primary use case is that you want to ignore empty directories, but what it in fact does is ignore _all_ directories.

marcphilipp commented:
I tend to agree with @pioterj on this one: `@IgnoreDirectories` sounds like directories (including their contents) are ignored.

pioterj commented:
I think it's totally fine if @IgnoreEmptyDirectories also ignored empty subdirectories. Unlike the current name, I'd not find it surprising or confusing at all.

lptr commented:
I'm also with @pioterj on using `@IgnoreEmptyDirectories`. We can clarify in documentation (Javadoc and user guide) as to how it handles hierarchies of otherwise empty directories. I don't think this will be confusing in practice.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputFilePropertyRegistration.java:
```diff
@@ -89,6 +92,23 @@ public TaskInputFilePropertyBuilderInternal withNormalizer(Class<? extends FileN
+
+    @Override
+    public TaskInputFilePropertyBuilder ignoreDirectories(boolean ingoreDirectories) {
+        this.directorySensitivity = ingoreDirectories ? DirectorySensitivity.IGNORE_DIRECTORIES : DirectorySensitivity.FINGERPRINT_DIRECTORIES;
```
```suggestion
        this.directorySensitivity = ignoreDirectories ? DirectorySensitivity.IGNORE_DIRECTORIES : DirectorySensitivity.FINGERPRINT_DIRECTORIES;
```

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java:
```diff
@@ -204,7 +206,7 @@ public void visitInputProperty(String propertyName, PropertyValue value, boolean
 
             @Override
-            public void visitInputFileProperty(final String propertyName, boolean optional, boolean skipWhenEmpty, boolean incremental, @Nullable Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
+            public void visitInputFileProperty(final String propertyName, boolean optional, boolean skipWhenEmpty, DirectorySensitivity directorySensitivity, boolean incremental, @Nullable Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
```
I think it would be nice to split these parameter declarations to separate lines now.

ghale replied:
Agreed.  I've cleaned this up in the various implementors.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskPropertyUtils.java:
```diff
@@ -21,13 +21,15 @@
      * Visits both properties declared via annotations on the properties of the task type as well as
      * properties declared via the runtime API ({@link org.gradle.api.tasks.TaskInputs} etc.).
      */
+    @UsedByScanPlugin("test-distribution")
```
We've talked briefly about this with @britter, @wolfs and @jonnybbb. I think it would be nice to find a better way to declare this dependency, i.e. by introducing a new subproject that we keep backwards compatible, like we do with the `:enterprise` subproject.

Not something we need to fix in this PR, but it's a good example to start the discussions from.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java:
```diff
@@ -320,7 +322,8 @@ public void visitInputs(InputVisitor visitor) {
                 String propertyName = inputFileProperty.getPropertyName();
                 visitor.visitInputFileProperty(propertyName, type, NON_IDENTITY, value, () -> {
-                    FileCollectionFingerprinter fingerprinter = fingerprinterRegistry.getFingerprinter(inputFileProperty.getNormalizer());
+                    FileCollectionFingerprinterSpec fingerprinterSpec = DefaultFileCollectionFingerprinterSpec.from(inputFileProperty.getNormalizer(), inputFileProperty.getDirectorySensitivity());
```
I wonder if we need the indirection of creating a `DefaultFileCollectionFingerprinterSpec` or if we should just pass in the `InputFilePropertySpec`.

Perhaps we could allow `InputFilePropertySpec` to implement `FileNormalizationSpec`.

ghale replied:
Good idea of making `InputFilePropertySpec` implement `FingerprinterSpec` (I assume that's what you meant by `FileNormalizationSpec`).

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/ModifierAnnotationCategory.java:
```diff
@@ -42,6 +43,9 @@
+    )),
+    DIRECTORY_ONLY(ImmutableSet.of(
+        IgnoreDirectories.class
     ));
```
I think this should either be called `IGNORE_EMPTY_DIRECTORIES` or if we want to be more general like the other categories, then `FILTERING`.

ghale replied:
Done.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputFilePropertyType.java:
```diff
@@ -32,6 +34,7 @@
         this.validationAction = validationAction;
     }
 
+    @UsedByScanPlugin("test-distribution")
```
This is not good. We shouldn't have these dependencies. /cc @britter @jonnybbb 

Not for this PR to fix.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputDirectoryPropertyAnnotationHandler.java:
```diff
@@ -34,7 +35,7 @@
     @Override
     public ImmutableSet<? extends AnnotationCategory> getAllowedModifiers() {
-        return ImmutableSet.of(INCREMENTAL, NORMALIZATION, OPTIONAL);
+        return ImmutableSet.of(INCREMENTAL, NORMALIZATION, OPTIONAL, DIRECTORY_ONLY);
```
This could use `ModifierAnnotationCategory.values()` instead, no?

lptr replied:
OK, I'm convinced.

lptr commented on subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/DirectorySensitivity.java:
```diff
@@ -0,0 +1,29 @@
+    // Ignore empty directories
+    IGNORE_DIRECTORIES,
+    // Not important (i.e. the fingerprinter does not handle empty directories one way or the other)
+    NONE
```
I suspect that if we removed `NONE` in favor of `DEFAULT`, then we could do away with the whole `DirectorySensitiveNormalizer` concept, too.

ghale replied:
Actually, thinking through this, and with the inclusion of absolute path fingerprinting in the directory normalization, this gets a lot simpler.  There are really only two types of fingerprinters now:

- Those that can be configured to either ignore or fingerprint directories (AbsolutePath, RelativePath, NameOnly) and default to fingerprinting directories.
- Those that do not fingerprint directories at all (Classpath, CompileClasspath, IgnoredPath) and cannot be configured.

`DEFAULT` feels a little squirrelly, because the "default" behavior is different between these two classes (the default for the first class is to fingerprint directories, the default, and only mode, for the second class is to exclude directories) but it does provide the nice advantage that everything is just "whatever the default is" unless you set it explicitly to ignore directories.

I've changed this to `DEFAULT` and `IGNORE_DIRECTORIES`.  I don't think we can get away without `DirectorySensitiveNormalizer` unless we produce an error if the user says something like  `ignoreDirectories().withNormalizer(ClasspathNormalizer)`.  This would search the fingerprinter registry for a fingerprinter registered with the `ClasspathNormalizer` and `IGNORE_DIRECTORIES` which won't exist.  It doesn't feel right to me to error in such a case, since ignoring directories is "valid" behavior for this class of normalizer.  To recognize that ignoring directories _is_ the default behavior and make sure that we search for a fingerprinter registered with `DEFAULT` means we would have to be able to identify which normalizers fall into this category (which would require something like `DirectorySensitiveNormalizer`).

I think the only way to get rid of `DirectorySensitiveNormalizer` would be to register a fingerprinter for both `DEFAULT` and `IGNORE_DIRECTORIES` for those where the default behavior is to ignore or to build some sort of knowledge of this distinction into the registry lookup.  These both feel worse to me.

I've changed this to `DEFAULT` vs `IGNORE_DIRECTORIES` but left `DirectorySensitiveNormalizer` in place, unless you have a better idea that I'm not thinking of.

lptr commented on subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/DefaultEmptySourceTaskSkipperTest.groovy:
```diff
@@ -42,7 +43,7 @@ class DefaultEmptySourceTaskSkipperTest extends Specification {
     final skipper = new DefaultEmptySourceTaskSkipper(cleanupRegistry, TestFiles.deleter(), outputChangeListener, taskInputsListeners)
     final fileCollectionSnapshotter = TestFiles.fileCollectionSnapshotter()
-    final fingerprinter = new AbsolutePathFileCollectionFingerprinter(fileCollectionSnapshotter)
+    final fingerprinter = new AbsolutePathFileCollectionFingerprinter(DirectorySensitivity.FINGERPRINT_DIRECTORIES, fileCollectionSnapshotter)
```
I expected `SourceTask` to acquire the new annotation. I think it will have some significant implications to backwards compatibility that in turn can have consequences in how we want to implement this. For that reason it would be good to explore those effects in this PR.

ghale replied:
I'd personally rather explore that in a follow up PR.  I think this PR provides useful functionality on its own and doesn't drag along the baggage of potentially breaking bits.  I'd rather have that plumbing in place first and then confront the problem of `SourceTask` and what we do about it.  If you feel strongly about it, I'm happy to be convinced otherwise, though.

lptr reviewed (changes requested):
Looks good so far, I left a bunch of comments. Most are centered around naming, and maybe some small potential code simplifications.

I have one major question, which is the case of `SourceTask`. I think we should adopt the new annotation for it in this PR to understand what it will mean in the long run.

britter commented:
Just another idea: How about reversing the naming and calling it `@FilesOnly` or something like that? Would that be clearer?

ghale commented:
I'm personally not a fan of `@FilesOnly` as it seems a little less clear to me what it actually does.  I'm fine with `@IgnoreEmptyDirectories` as it addresses the primary (maybe only) use case that people will use this and seems very clear to me.  As long as everyone is comfortable with the fact that it actually does a little more than _just_ ignoring empty directories (and it seems like we are) then I'm +1 for it, too.

ghale commented on subprojects/core/src/main/java/org/gradle/internal/fingerprint/FileCollectionFingerprinterSpec.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Specifies criteria for selecting a {@link FileCollectionFingerprinter}
+ */
+public interface FileCollectionFingerprinterSpec {
```
+1 Changed.

ghale commented on subprojects/core-api/src/main/java/org/gradle/api/tasks/IgnoreDirectories.java:
```diff
@@ -0,0 +1,47 @@
+import java.lang.annotation.Target;
+
+/**
+ * Attached to a task input property to specify that directories should be ignored
```
I've just updated this to "input property" instead of "task input property".

ghale commented on subprojects/core-api/src/main/java/org/gradle/api/tasks/IgnoreDirectories.java:
```diff
@@ -0,0 +1,47 @@
+ *
+ * <ul><li>{@link org.gradle.api.tasks.InputFiles}</li>
+ *
+ * <li>{@link org.gradle.api.tasks.InputDirectory}</li> </ul>
```
This comes from modeling it after the `SkipWhenEmpty` javadoc.  I've changed it to explain that these are the types it can be applied to as well as to add the transform property types.

ghale commented on subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java:
```diff
@@ -36,15 +39,18 @@
-    public static final FingerprintingStrategy IGNORE_MISSING = new AbsolutePathFingerprintingStrategy(false);
+    public static final FingerprintingStrategy INCLUDE_MISSING = new AbsolutePathFingerprintingStrategy(true, DirectorySensitivity.FINGERPRINT_DIRECTORIES);
+    public static final FingerprintingStrategy IGNORE_MISSING = new AbsolutePathFingerprintingStrategy(false, DirectorySensitivity.FINGERPRINT_DIRECTORIES);
+    public static final FingerprintingStrategy IGNORE_DIRECTORIES = new AbsolutePathFingerprintingStrategy(true, DirectorySensitivity.IGNORE_DIRECTORIES);
```
Thanks for the heads up.  I see this has been merged and I've rebased on top of it.  I believe all of the conflicts have been resolved properly.

ghale commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/DirectorySensitivityErrorHandlingIntegrationSpec.groovy:
```diff
@@ -0,0 +1,93 @@
+
+    enum NonDirectoryInput {
+        INPUTFILE("@${InputFile.class.simpleName}", 'File', 'file("foo")', "@${PathSensitive.class.simpleName}(${PathSensitivity.class.simpleName}.${PathSensitivity.RELATIVE.name()})"),
+        INPUT("@${Input.class.simpleName}", 'String', '"foo"')
```
True.  Done.

ghale commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/DirectorySensitivityErrorHandlingIntegrationSpec.groovy:
```diff
@@ -0,0 +1,93 @@
+        succeeds("taskWithInputs")
+
+        where:
+        nonDirectoryInput << NonDirectoryInput.values()
```
Good point.  Done.

lptr commented on subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputFilePropertyBuilder.java:
```diff
@@ -79,14 +79,4 @@
-     * @since 6.8
-     */
-    @Incubating
-    TaskInputFilePropertyBuilder ignoreEmptyDirectories(boolean ignoreEmptyDirectories);
```
Why remove this? We have the same for `optional()` and `skipWhenEmpty()`.

ghale replied:
Yeah, that's the only reason I had put it in there in the first place - because it was there for other things.  I removed it because we changed `DirectorySensitivity` to be either `DEFAULT` or `IGNORE`, so it no longer made sense.  We no longer have something that represents that you explicitly want to fingerprint directories (i.e. `ignoreEmptyDirectories(false)`).

For instance, if someone did something like `ignoreDirectories(false).withNormalizer(CLASSPATH)` what should we do?  Ideally we would throw an error, but how do we tell when the value can only be `DEFAULT` or `IGNORE`?  So we could either silently allow nonsensical behavior, which seems wrong; or we could just track the fact that you tried to set `ignoreDirectories()` at all with a non-configurable normalizer, which means we would error on `ignoreDirectories(true)` as well, which also seems wrong; or we could re-introduce the "I explicitly set it to fingerprint directories" state (which we just removed); or we just remove this method which doesn't make complete sense anyways (i.e. there is no normalizer where it would be _necessary_ to say `ignoreEmptyDirectories(false)`).

ghale replied:
Another option here, I suppose would be to expose `DirectorySensitvity` on the public api and allow users to set it that way, but that feels kind of gross, too.

lptr replied:
Hm. I don't think we've been very rigorous with this API in the past, e.g. `skipWhenEmpty(true)` can be applied to a `file()` input where we don't actually take it into account.

To me calling `ignoreDirectories(false)` _could_ mean not to specifically ignore directories (i.e. `DEFAULT`).

As to `ignoreDirectories(false).withNormalizer(CLASSPATH)` our current rule is to use the latest specified values. We allow for example to do `withPathSensitivity(NONE).withNormalizer(CLASSPATH).withPathSensitivity(ABSOLUTE)`.

So I'd just treat this as a mutable thing where calling one method might override the result of a previous call.

ghale replied:
Ok, maybe I'm being too strict about not allowing thie sort of meaningless configuration.  I've re-introduced the `ignoreEmptyDirectories(boolean)` method.  If someone were to say `ignoreDirectories(false).withNormalizer(CLASSPATH)`, it will just accept this and operate as normal.

lptr commented on subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/DirectorySensitivity.java:
```diff
@@ -0,0 +1,28 @@
+    // default behavior is to fingerprint directories, for others, they ignore directories by default.
+    DEFAULT,
+    // Ignore empty directories
+    IGNORE_DIRECTORIES
```
Why not use Javadoc for these?

ghale replied:
Heh, no reason...fixed.

lptr commented on subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintController.kt:
```diff
@@ -51,13 +54,15 @@ class ConfigurationCacheFingerprintController internal constructor(
     private val directoryFileTreeFactory: DirectoryFileTreeFactory
 ) : Stoppable {
+    private
+    val fileCollectionFingerprinter = fingerprinterRegistry.getFingerprinter(DefaultFingerprinterSpec.from(AbsolutePathInputNormalizer::class.java, DirectorySensitivity.DEFAULT))
```
It would be good to avoid injecting the whole registry here. I think it should be possible to create a service-provider method that passes in the single `FileCollectionFingerprinter` here instead.

ghale replied:
I'm not sure if what I did was what you were expecting or if there is some obscure feature of `ServiceRegistry` you wanted me to use.  Basically, the issue here is that looking up just an `AbsolutePathFileCollectionFingerprinter` doesn't work, because there are multiple registered objects of that type (`FileCollectionFingerprinterRegistry` grabs all of those and stuffs them in the registry).  We might want to think about registering the instances in the fingerprinter registry differently so we only have to provide service methods for the fingerprinter(s) that we need to get explicitly.  

If you had something else in mind, let me know.

lptr replied:
OK. Let's just leave this as it is now.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/fingerprint/FingerprinterSpec.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Specifies criteria for selecting a {@link FileCollectionFingerprinter}
+ */
+public interface FingerprinterSpec {
```
I think this would better be called `FileNormalizationSpec`. We currently only use "fingerprinting" for input _files,_ but this can change if we introduce normalization for input values.

I also dislike "fingerprinting" as a name, and would love to replace it with something else that is more different to "snapshotting." Thus it would be nice not to embed the word more in the codebase.

ghale replied:
Done

lptr commented on subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultFingerprinterSpec.java:
```diff
@@ -0,0 +1,70 @@
+        } else {
+            return new DefaultFingerprinterSpec(normalizer, DirectorySensitivity.DEFAULT);
+        }
+    }
```
I think if we wanted to check something here, we should throw errors if the parameters are incorrect.

ghale replied:
At the moment, the parameters can't be incorrect.  For `DirectorySensitiveNormalizer`s the value is either `DEFAULT` (fingerprint directories) or `IGNORE_DIRECTORIES`, both of which are valid and represent different instances in the registry.  For other normalizers, the value is either `DEFAULT` (ignore directories) or `IGNORE_DIRECTORIES`, which are the same intention.  All we're doing here is making sure that for these other normalizers, we always look for the right thing in the registry (i.e. a normalizer with `DEFAULT` directory sensitivity.  See my comments here: https://github.com/gradle/gradle/pull/15147#discussion_r526416209

lptr commented on subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/properties/DefaultPropertyWalkerTest.groovy:
```diff
@@ -61,11 +61,11 @@ class DefaultPropertyWalkerTest extends AbstractProjectBuilderSpec {
         1 * visitor.visitInputProperty('bean', { it.call() == NestedBean }, false)
         1 * visitor.visitInputProperty('bean.nestedInput', { it.call() == 'nested' }, false)
-        1 * visitor.visitInputFileProperty('bean.inputDir', _, _, _, _, _, InputFilePropertyType.DIRECTORY)
+        1 * visitor.visitInputFileProperty('bean.inputDir', _, _, _, _, _, _, InputFilePropertyType.DIRECTORY)
```
Gah! Not for this PR, but I wonder how this could be made nicer.

lptr commented on subprojects/snapshots/src/test/groovy/org/gradle/internal/fingerprint/impl/PathNormalizationStrategyTest.groovy:
```diff
@@ -92,7 +93,7 @@ class PathNormalizationStrategyTest extends Specification {
 
     def "sensitivity NAME_ONLY"() {
-        def fingerprints = collectFingerprints(NameOnlyFingerprintingStrategy.INSTANCE)
+        def fingerprints = collectFingerprints(NameOnlyFingerprintingStrategy.FINGERPRINT_DIRECTORIES)
```
I'm not completely sure, but to me `DEFAULT` sounds like a better name here, too.

ghale replied:
Changed.

lptr commented on subprojects/snapshots/src/test/groovy/org/gradle/internal/fingerprint/impl/PathNormalizationStrategyTest.groovy:
```diff
@@ -92,7 +93,7 @@ class PathNormalizationStrategyTest extends Specification {
     }
 
     def "sensitivity NAME_ONLY"() {
```
Should this test class include tests for the include/exclude empty directories cases?

ghale replied:
Yeah probably.  Done.

wolfs commented:
@ghale I created https://github.com/gradle/gradle/pull/15256 on top of this PR to see if adding the new annotation to source task breaks anything. It looks pretty good for now! So we probably should integrate the changes from my PR into this PR.

lptr commented:
@gary can you please cherry-pick @wolfs's changes to this PR? Or do you want to apply the annotation on `SourceTask` separately?

lptr commented on subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java:
```diff
@@ -411,8 +412,13 @@ OutputSnapshotter createOutputSnapshotter(FileCollectionSnapshotter fileCollecti
+        Supplier<AbsolutePathFileCollectionFingerprinter> createDefaultAbsoluteFileCollectionFingerprinterSupplier(FileCollectionSnapshotter fileCollectionSnapshotter) {
+            AbsolutePathFileCollectionFingerprinter defaultAbsolutePathFileCollectionFingerprinter = new AbsolutePathFileCollectionFingerprinter(DirectorySensitivity.DEFAULT, fileCollectionSnapshotter);
+            return () -> defaultAbsolutePathFileCollectionFingerprinter;
+        }
```
I don't think this is making things better, let's revert this and pass the registry for now. We can revisit this later I think.

ghale replied:
Agreed.  Done.

lptr reviewed (changes requested):
I think we should revert to passing the whole registry instead of the `Supplier`-based solution.

I also think we should add back the `boolean`-accepting variant of the runtime API method, but I can be convinced otherwise.

And I think we should add @wolfs's commits to enable the new feature for `SourceTask` and friends.

Apart from these I think we are good to go.

wolfs commented on subprojects/core-api/src/main/java/org/gradle/api/tasks/IgnoreEmptyDirectories.java:
```diff
@@ -0,0 +1,54 @@
+
+/**
+ * Attached to an input property to specify that directories should be ignored
+ * when snapshotting inputs.  Files within directories and subdirectories will be
```
```suggestion
 * when snapshotting inputs. Files within directories and subdirectories will be
```

Nit.

wolfs commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformationRegistrationFactory.java:
```diff
@@ -201,10 +208,19 @@ public TransformationStep getTransformationStep() {
 
         @Override
-        public void visitInputFileProperty(String propertyName, boolean optional, boolean skipWhenEmpty, boolean incremental, @Nullable Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
+        public void visitInputFileProperty(String propertyName,
+                                           boolean optional,
+                                           boolean skipWhenEmpty,
+                                           DirectorySensitivity directorySensitivity,
+                                           boolean incremental,
+                                           @Nullable Class<? extends FileNormalizer> fileNormalizer,
+                                           PropertyValue value,
+                                           InputFilePropertyType filePropertyType) {
```
```suggestion
        public void visitInputFileProperty(
            String propertyName,
            boolean optional,
            boolean skipWhenEmpty,
            DirectorySensitivity directorySensitivity,
            boolean incremental,
            @Nullable Class<? extends FileNormalizer> fileNormalizer,
            PropertyValue value,
            InputFilePropertyType filePropertyType
        ) {
```

Nit: I think it looks much nicer when the first argument and the closing bracket is on a new line.

There are a few more method calls/constructors where the first argument is on the same line as the opening bracket, I didn't comment on any of those.

wolfs commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/AbstractDirectorySensitivityIntegrationSpec.groovy:
```diff
@@ -0,0 +1,535 @@
+        [api, pathSensitivity] << [API.values(), [PathSensitivity.RELATIVE, PathSensitivity.ABSOLUTE, PathSensitivity.NAME_ONLY]].combinations()
+    }
+
+    @ToBeFixedForConfigurationCache(because = "doesn't like FileCollection in transform parameters")
```
That is very strange.  An abstract getter for `ConfigurableFileCollection` should work perfectly well for transform parameters.

wolfs commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -90,6 +90,38 @@ This improves the likelihood of up-to-date and build cache hits when any propert
+
+For [up-to-date checks](userguide/more_about_tasks.html#sec:up_to_date_checks) and the [build cache](userguide/build_cache.html), Gradle needs to determine if two directory structures contain the same contents.  When a directory contains an empty directory, it is considered to have different contents than an identical directory where the empty directory does not exist.
+
+This may not always be desirable.  There are many cases where only the files in a directory structure may be significant, and an empty directory will have no impact on the outputs of a task.  In such cases, re-executing the task because an empty directory exists is unnecessary as it will only produce the same outputs.
```
```suggestion
This may not always be desirable. There are many cases where only the files in a directory structure may be significant, and an empty directory will have no impact on the outputs of a task. In such cases, re-executing the task because an empty directory exists is unnecessary as it will only produce the same outputs.
```

Nit: extra space

wolfs commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -90,6 +90,38 @@ This improves the likelihood of up-to-date and build cache hits when any propert
+
+This may not always be desirable.  There are many cases where only the files in a directory structure may be significant, and an empty directory will have no impact on the outputs of a task.  In such cases, re-executing the task because an empty directory exists is unnecessary as it will only produce the same outputs.
+
+A new annotation has been introduced to address this scenario.  Inputs annotated with [@InputFiles](javadoc/org/gradle/api/tasks/InputFiles.html) or [@InputDirectory](javadoc/org/gradle/api/tasks/InputDirectory.html) can additionally be annotated with [@IgnoreDirectories](javadoc/org/gradle/api/tasks/IgnoreDirectories.html) to specify that directories should not be considered during build cache a ...
```
```suggestion
A new annotation has been introduced to address this scenario. Inputs annotated with [@InputFiles](javadoc/org/gradle/api/tasks/InputFiles.html) or [@InputDirectory](javadoc/org/gradle/api/tasks/InputDirectory.html) can additionally be annotated with [@IgnoreDirectories](javadoc/org/gradle/api/tasks/IgnoreDirectories.html) to specify that directories should not be considered during build cache and up-to-date checks.
```

Nit: extra space

wolfs commented on subprojects/docs/src/docs/userguide/authoring-builds/more_about_tasks.adoc:
```diff
@@ -655,6 +655,10 @@ Implies `<<#incremental,@Incremental>>`.
+
+| `@link:{javadocPath}/org/gradle/api/tasks/IgnoreEmptyDirectories.html[IgnoreEmptyDirectories]`
+| `Iterable&lt;File&gt;`*
+| Used with `@InputFiles` or `@InputDirectory` to instruct Gradle to track only changes to the contents of directories and not differences in the directories themselves.  For example, removing, renaming or adding an empty directory somewhere in the directory structure will not make the task out-of-date.
```
```suggestion
| Used with `@InputFiles` or `@InputDirectory` to instruct Gradle to track only changes to the contents of directories and not differences in the directories themselves. For example, removing, renaming or adding an empty directory somewhere in the directory structure will not make the task out-of-date.
```

Nit: extra space

Shouldn't we use one line per sentence in the user guide? I see that the other elements of this table don't do this, though I wonder if we at least should use the style for the new content.

wolfs commented on subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/AbsolutePathFingerprintingStrategy.java:
```diff
@@ -52,18 +59,26 @@ public String normalizePath(FileSystemLocationSnapshot snapshot) {
     }
 
+    private boolean shouldFingerprint(CompleteFileSystemLocationSnapshot snapshot) {
+        return !(snapshot.getType() == FileType.Directory && directorySensitivity == DirectorySensitivity.IGNORE_DIRECTORIES);
```
Can we move this method to the `DirectorySensitivity` enum? The implementation could even be per enum constant.

ghale replied:
Good call.  I really like the way this worked out.  Much better now.

wolfs reviewed (commented):
I left some comments.

+1 for adding back the boolean-accepting variant of the runtime API method.

None of my comments blocks merging this and they can be addressed in a follow up commit for e.g. 6.9/7.0.

ghale commented:
@lptr I've implemented the changes that you wanted\, including merging the `SourceTask` changes.

@wolfs I've implemented a few of the changes that you suggested, but some of the more cosmetic changes I'd like to leave for follow up work, per your comments.

Assuming this clears CI, are we good to merge now?

wolfs commented on subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/DirectorySensitivity.java:
```diff
@@ -0,0 +1,47 @@
+
+    private final Function<FileSystemLocationSnapshot, Boolean> fingerprintCheck;
+
+    DirectorySensitivity(Function<FileSystemLocationSnapshot, Boolean> fingerprintCheck) {
```
This should probably be a predicate.

ghale replied:
🤦  Yes, yes it should...

wolfs reviewed (approved):
Lgtm! Let's merge it if CI passes and do the remaining things in a follow-up PR. We also need to add upgrade notes about the possibly breaking source task changes.

ghale commented:
@wolfs Just fyi - I did add upgrade notes about the potential breaking behavior when I cherry-picked your changes.
